hobo/src/storage.rs

```diff
@@ -151,34 +154,91 @@ impl<Component: 'static> Storage<Component> for SimpleStorage<Component> {
 		self.data.get_mut(&entity).unwrap()
 	}
 }
+// add location make real struct
+pub struct StorageGuard<Component: 'static, Inner: std::ops::Deref<Target = SimpleStorage<Component>>> {
+	pub inner: Option<Inner>,
+	#[cfg(debug_assertions)]
+	pub location: std::panic::Location<'static>,
+}
+pub struct StorageGuardMut<'a, Component: 'static, Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>> {
+	pub world: &'a World,
+	pub inner: Option<Inner>,
+	#[cfg(debug_assertions)]
+	pub location: std::panic::Location<'static>,
+}
+
+unsafe impl<Component: 'static, Inner: std::ops::Deref<Target = SimpleStorage<Component>>> owning_ref::StableAddress for StorageGuard<Component, Inner> {}
+unsafe impl<'a, Component: 'static, Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>> owning_ref::StableAddress for StorageGuardMut<'a, Component, Inner> {}
+
+impl<Component, Inner> std::ops::Deref for StorageGuard<Component, Inner> where
+	Component: 'static,
+	Inner: std::ops::Deref<Target = SimpleStorage<Component>>,
+{
+	type Target = SimpleStorage<Component>;
 
-pub struct StorageGuard<'a, Component: 'static, Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>>(pub &'a World, pub Option<Inner>);
-unsafe impl<'a, Component: 'static, Inner: std::ops::DerefMut<Target = SimpleStorage<Component>>> owning_ref::StableAddress for StorageGuard<'a, Component, Inner> {}
+	fn deref(&self) -> &Self::Target { self.inner.as_ref().unwrap() }
+}
+
+#[cfg(debug_assertions)]
+impl<Component, Inner> Drop for StorageGuard<Component, Inner> where
+	Component: 'static,
+	Inner: std::ops::Deref<Target = SimpleStorage<Component>>,
+{
+	fn drop(&mut self) {
+		let StorageGuard { location, .. } = self;
+		crate::backtrace::STORAGE_MAP.0.borrow_mut()
+			.entry(std::any::TypeId::of::<Component>())
+			.and_modify(|map| { map.remove(location); });
```

Comment:
you should check that there's no mutable borrows in existence and panic if there are any, same as when dropping the mutable one except you're interested not in the number of borrows but in whether any are mutable